### ISSUE TITLE
FNX-18 ⁃ [Spike] Holding back does not show tab stack history

### ISSUE THREAD

**Issue by yoasif** [mozilla-mobile/fenix#1048]
### Acceptance Criteria
* [ ] Make an MVP prototype where user can jump to a specific spot in tab history (may involve adding the back button back) or have some recommendations about what eng restrictions there are
* [ ] timebox this to 2 days
* [ ] Use a feature flag (and don't land any strings until 7/18)

## Steps to reproduce
1. Open a page with links
1. Click a link
1. Click a link
1. Hold back button

### Expected behavior
Menu appears with tab history, allowing me to navigate back to the page in step 1 above. 

### Actual behavior
Holding back does nothing. 

### Device information
* Android device: Pixel 2
* Fenix version: 1.0.1912 (Build #10760203 :lizard: 67.0a1-20190312095443)

**Comment by cpeterso** [comment]
I think this history UI will need the GV "history stack" API that should land before the end of M3 (March): https://bugzilla.mozilla.org/show_bug.cgi?id=1509110

**Comment by Amejia481** [comment]
This is what happens when you hold press back button on Fennec
![Screenshot_20190319-134517](https://user-images.githubusercontent.com/773158/54629321-b91af680-4a4d-11e9-9d28-2e94f7ffbabb.png)

**Comment by gurpreetatwal** [comment]
Is this feature still planned to be included?

**Comment by NotWoods** [comment]
It's in the backlog

**Comment by danielzgtg** [comment]
@kbrosnan This won't work with Android 10 gesture navigation. The system considers a held back gesture as the same thing as a normal back.

https://github.com/mozilla-mobile/fenix/issues/8083#issuecomment-590135738

> How about we make long-press forward show both forward and back tab history?

**Comment by buttercookie42** [comment]
Given that Chrome only relatively recently implemented the "show tab history when long-pressing the back button" behaviour, what do _they_ do in that case?
And can't the classic back button behaviour still be implemented in any case when gesture navigation isn't enabled (or straight out not available on older devices)?

**Comment by danielzgtg** [comment]
It doesn't work in Chrome. They still need to add tab history for gesture navigation.

Yes, the classic back button behavior can and should be implemented. I meant “This won't work […]” to be in response to the other issue being closed.

**Comment by madb1lly** [comment]
Since I've been using Fenix Nightly as a daily driver I find the lack of this feature frustrating (and the complementary long press on the forward icon in the menu).

**Comment by madb1lly** [comment]
Since I've been using Fenix Nightly as a daily driver I find the absence of this Fennec feature frustrating (and the complementary long press on the forward icon in the menu).

**Comment by Poopooracoocoo** [comment]
Focus issue: https://github.com/mozilla-mobile/focus-android/issues/3845

**Comment by robsmith11** [comment]
Will the back history be added before Fenix replaces Fennec as the stable release?

It's a pretty significant functionality feature that is missing and doesn't seem to be on any milestone lists.

**Comment by kevincox** [comment]
Yup, this is a critical feature when you get websites that loop on back. Without a way to jump multiple pages at a time thee is no way to escape other than closing the tab.

**Comment by liuche** [comment]
This is a feature that a lot of Fennec Beta migrated users have been asking for, and @vesta0 would like eng to do a little timeboxed exploration/thinking about how this would be possible. We will need UX at some point (since we don't even have a back button anymore) but coming up with some rough ideas.

**Comment by robsmith11** [comment]
For me, and I would guess many others, the way Fennec implemented it was perfect.  Just re-add the back button to the main menu and show the history chain on a long press (ideally should work for forward button too).

There's no need for it to work with gestures.  The "back swipe" can still just trigger a 1-step move back in history.

**Comment by yoasif** [comment]
> We will need UX at some point (since we don't even have a back button anymore) but coming up with some rough ideas.

FWIW, I still see a back button -- it isn't on a toolbar or in a menu, but I never used the one in the toolbar in Fennec anyway -- I just tried it there, and it shows the history stack at the bottom of the screen, while the back toolbar button is at the top of the screen - which feels pretty unintuitive and unexpected (I can better understand why people want to see https://github.com/mozilla-mobile/fenix/issues/7461).

I continue to use the non-gesture based back button in Android 10 (it also still works in Fennec).

**Comment by Poopooracoocoo** [comment]
What a lot of people have suggested is tapping and holding on the forward button.

**Comment by madb1lly** [comment]
Hi @yoasif are you talking about Fennec or Fenix? Your description sounds like Fennec, because in Fenix I can't see a back toolbar button at the top of the screen in Fenix even if I set the toolbar to be at the top. I cannot get a tab history stack to appear anywhere in Fenix.

I personally don't see the issue in re-adding the back button the menu. There are several good reasons to do this including accessibility #7335 and this tab history function. There's plenty of white space in that menu bar, why not relax some arbitrary UX rules and put the back button back.

Cheers 🙂

**Comment by yoasif** [comment]
@madb1lly I'm talking about the [Android back button](https://developer.android.com/guide/navigation/navigation-custom-back).

@Poopooracoocoo That is a weird solution, imo - on Fennec, holding down forward shows the *forward* stack, not the back stack. So it'd be a weird transition and pretty intuitive (imo) to show things in *front* of what I am looking at by selecting *back*.

To the UX folks - I think this is tough on the newer gesture based UIs, but I wouldn't want to throw out the baby with the bathwater on the stack idea for non-gesture based devices - not only do older devices not have gestures, people with newer devices don't always use them.

I don't actually think that the back button on Fennec's toolbar + the old tab stack was the best design, since the stack/menu appeared far away from where I am holding down my finger. 

Don't really have any great ideas on the challenges around gestures - part of the reason I quickly disabled it on my device after trying it yesterday was that a forward swipe (vs. a back swipe) *also* goes back, and that just feels inconsistent. That is more of an issue for Android, and I am sure they have great reasons for it.

Some quick observations:

Edge, Opera and Samsung Internet have a bottom toolbar with history stacks in both back and forward buttons. Edge and Samsung Internet's autohides.

Kiwi and Brave hide their forward button (without a tab stack) inside of their main meatball menu. 

Opera Touch includes back and forward buttons inside of their main meatball menu. It does not feature a tab stack.

**Comment by buttercookie42** [comment]
Chrome also has a tab history popup when long-pressing the (system) back button. Long-pressing the forward button in the menu does nothing, and I've no idea how it works on devices with gesture navigation. Interestingly their popup only shows the *back* history, though I'm definitively not suggesting to copy *that*.

**Comment by danielzgtg** [comment]
@yoasif It was me that suggested this in #8083.

My reason for suggesting it be put on the forward button is because it was the closest thing that had to do with the back stack after other people declined to readd the back button.

It is not weird to show the back stack after long pressing the forward button. Back, current, and forward are as related as past, present, and future. There shouldn't even be a distinction between back and forward stacks. There should be just one single history timeline.

Currently, if I really need to open the back stack in Chrome, I have to go into Android Settings and turn off gesture navigation. Unfortunately, Chrome Dev has removed the functionality completly and now the situation there is as bad as Fenix.

In comparison to Chrome and Fennec, Fenix is able to open neither back nor forward stacks. Being able to access this functionality one way or another would at least be a step up. It doesn't matter if for long press forward we filter the history for only forward entries right now. My suggestion for enhancing the functionality to support Android 10 gesture navigation by including both back and forward in a single history timeline can wait.

**Comment by Poopooracoocoo** [comment]
@danielzgtg oh shoot. as always, google sucks

**Comment by Matth7878** [comment]
If long pressing the forward is used to show back history it should also displayed forward history.
If you went back several times in history you might want to go back forward several times.

I would suggest that long pressing forward button should display inside the tab all history of the tab. (And not in a kind of contextual menu like before)
So it would be easy to navigate your history either back or forward.

So you can freely choose

**Comment by Poopooracoocoo** [comment]
@Matth7878 I personally like the context menu thing at the bottom but if it became a screen or centred window when there's heaps of entries, then that's cool.

**Comment by madb1lly** [comment]
Thanks for clarifying @yoasif 🙂

I really don't mind the implementation so long as the functionality is available. I really don't understand what the problem is with using some of that white space for a back button though.

Cheers 🙂

**Comment by Lungenpest** [comment]
I really need the old behavior of the back button like in legacy Firefox for Android. That helps me to fast navigate between some discussion pages in seperate tabs. I don´t have any buttons on my android device, because I activated the One Plus gesture navigation on my OP6T A6013 (Android 10). Please take a look at this Video, there you can see what i need and maybe you can understand. It´s really hard to navigate this way in seperate tabs in fenix.

[https://youtu.be/wJgSyiLfa_o](url)

**Comment by liuche** [comment]
hey @brampitoyo it looks like this a fennec feature that people from the Migration are really missing - I know UX is super overloaded, so I'd like to let engineering do a proof of concept of this.

We can just match the Fennec mocks for now, and even keep it on Nightly, and see how it feels. How does that sound?

**Comment by brampitoyo** [comment]
@liuche I think this sounds ok.

Regardless of the result of the discussion below, opening “tab stack history” when holding the forward button makes sense. It’s an action that won’t impact or interfere with other features. The problem here is that the forward button isn’t always available to access.

@apbitner and @topotropic have looked into our main menu, so I want to first check in with them about re-adding the Back button.

### Question for @apbitner @topotropic 

The “tab stack history” feature seems to be a feature that users miss after using Fenix for some time.

1. Is there an argument to be made for reintroducing a software Back button, that when long-pressed will open a “tab stack history” page?
2. Do we have enough space inside the main menu to show this Back button alongside 4 other icons? What compromises might we have to make?
3. Does it make sense to design a totally separate page for this purpose (as in Fennec), or is there some way to repurpose our current History page? Currently, it shows all browsing history, but we could have a separate section called “History of current tab” or something.

### Question for @liuche and Engineering

1. We’ve removed the Back button from the main menu because Fenix prefers to use the Android system back button. Is it possible to trigger “tab stack history” by long-pressing the system button?
2. If the user runs Android 10, which doesn’t have dedicated back button, can we still trigger “tab stack history” by e.g. swiping left/right and holding the gesture for a bit?

Below, I’m posting some examples of “tab stack history” based on @yoasif’s https://github.com/mozilla-mobile/fenix/issues/1048#issuecomment-618456771 for reference:

#### Edge

<img width="360" src="https://user-images.githubusercontent.com/51007/81128779-3d07a900-8f96-11ea-8185-45ebfad28416.png">

#### Samsung Internet

<img width="360" src="https://user-images.githubusercontent.com/51007/81128788-4264f380-8f96-11ea-8481-92e88f939375.png">

#### Opera

<img width="360" src="https://user-images.githubusercontent.com/51007/81128848-7b04cd00-8f96-11ea-862e-5bd23f643503.png">

#### Opera Mini

![Screenshot_20200506-123229_Opera_Mini](https://user-images.githubusercontent.com/51007/81128858-8657f880-8f96-11ea-99d9-f87b5fb3eac5.png)

**Comment by kbrosnan** [comment]
### Question 1.  Is it possible to trigger “tab stack history” by long-pressing the system button?

Yes Fennec does this. Though we might need some thinking about how this behaves for users who only use gesture navigation. As that seems to be the direction Android project wants users to have as a default.

**Comment by yoasif** [comment]
>If the user runs Android 10, which doesn’t have dedicated back button, can we still trigger “tab stack history” by e.g. swiping left/right and holding the gesture for a bit?

If this is possible, it makes the most sense to me, although making it work alongside https://github.com/mozilla-mobile/fenix/issues/10240 might be weird if they *also* have Android gesture navigation enabled.

>Is there an argument to be made for reintroducing a software Back button, that when long-pressed will open a “tab stack history” page?

For most of the browsers that offer this feature, they mostly offer an in-app (vs. Android) back button -- as you noted, @brampitoyo:

* Edge
* Samsung Internet
* Opera
* Fennec

Samsung Internet seems to a true believer in gesture based navigation in Android, because they don't offer the stack on long press on the Android back button, while Edge, Opera and Fennec do. 

Kiwi is the only one I noticed that offers a stack on long press of the Android back button, but offers no back button in-app (and correspondingly, no stack) - and long pressing the forward button in their kebab menu does not offer a stack either. 

>Do we have enough space inside the main menu to show this Back button alongside 4 other icons? What compromises might we have to make?

Fennec makes a different call here -- it shows _back, forward, bookmark, and reload_ - _share_ is a menu item instead of an icon. I kinda feel like this order makes more sense. I will also note that Chrome also relegates _share_ to a menu item, although it has different icons in their menu. 

>Does it make sense to design a totally separate page for this purpose (as in Fennec), or is there some way to repurpose our current History page? Currently, it shows all browsing history, but we could have a separate section called “History of current tab” or something.

This seems like a bad idea, imo - history is simply not near the tab - it would take at least 3 taps on the screen to get to the Library UI, and likely a fourth to show the tab history. 

---

I do have a couple of wonky ideas on how this could work in a gesture based Android configuration:

1. Bring back some variation of the QAB. See also: https://github.com/mozilla-mobile/fenix/issues/7151

I was playing with Chrome to get a better feel of how gestures work there and in other browsers, and I noticed that pulling down the address bar on both Chrome and Edge brought me to the tab switcher. 

This gesture could also work for a QAB alike - UX could also build an affordance for items that lived within that area like the kebab menu that lives inside of the address bar on desktop (although it would probably be weird to have *two* kebab menus on the same toolbar -- perhaps the existing menu could be a hamburger menu instead - Samsung Internet uses this UI).

One of those items might be "History of current tab" as you suggested @brampitoyo 

![firefox-kebab-address](https://user-images.githubusercontent.com/152493/81132648-2d984980-8f1d-11ea-8ae7-23f671892e04.png)

That could also contain items like the reader mode icon - see https://github.com/mozilla-mobile/fenix/issues/6639

2. Introduce a history drop marker

This was very recently killed off in desktop, but adding a drop marker like used to exist prior to Firefox 74 on desktop would make it easy to have a UI affordance to allow users to view both backwards and forwards in their tab stack without needing to log press on an inconsistent button.

![firefox drop marker](https://user-images.githubusercontent.com/152493/81133026-608f0d00-8f1e-11ea-99b8-4bd6297a2c59.png)

It also makes it possible to not have to add *both* back and forward like Edge, Samsung Internet and Opera do -- although they clearly have an easier time of it as they have a dual pane toolbar layout (top and bottom), whereas Fenix clearly wants to maintain a slimmer one panel layout.

**Comment by cadeyrn** [comment]
> Is there an argument to be made for reintroducing a software Back button

It was already mentioned that - by default - there is no longer an Android back button since Android 10. I can't speak for others but for me it's not possible at all to use Fenix with one hand with gestures enabled if I want to navigate one page back. But I can reach the buttons in the menu very easy so this seems to be a very strong argument for users who want to use the browser with only one hand.

**Comment by yoasif** [comment]
> > Is there an argument to be made for reintroducing a software Back button
> 
> It was already mentioned that - by default - there is no longer an Android back button since Android 10. I can't speak for others but for me it's not possible at all to use Fenix with one hand with gestures enabled if I want to navigate one page back. And even the Android back button is difficult to reach for me (but possible at least). I can reach the buttons in the menu very easy so this seems to be a very strong argument for users who want to use the browser (mostly) with only one hand. Navigating one or more pages back is a feature I need very often every day. So a software back button would definitively _make my life easier_.

@cadeyrn It seems like you are advocating for a back button hidden inside of a menu, not an on screen one like Samsung Internet, Opera, or Edge - is that accurate? 

Is this preferable to one that exists inside of a menu?

**Comment by cadeyrn** [comment]
> @cadeyrn It seems like you are advocating for a back button hidden inside of a menu, not an on screen one like Samsung Internet, Opera, or Edge - is that accurate?

(based on screenshots from Samsung Internet, Opera and Edge, I don't have any of these browsers installed)

Sure, the placement of the back button in Samsung Internet, Opera or Edge is better than no back button at all for users on Android 10+ with gestures instead of the Android buttons.

But it's still difficult to reach for me because it's the most left button in all these browsers. It's the same as with the Android back button. With a placement in the menu it's one step more but the placement is a much bigger advantage for me than the additional step is a disadvantage for me.

**Comment by Poopooracoocoo** [comment]
> #### Question 1. Is it possible to trigger “tab stack history” by long-pressing the system button?
> Yes Fennec does this. Though we might need some thinking about how this behaves for users who only use gesture navigation. As that seems to be the direction Android project wants users to have as a default.

@kbrosnan is it possible to detect when a user swipes and holds (for the back gesture)?

**Comment by liuche** [comment]
How much engineering work is involved? Will add deliverables.

**Comment by yoasif** [comment]
> If the user runs Android 10, which doesn’t have dedicated back button, can we still trigger “tab stack history” by e.g. swiping left/right and holding the gesture for a bit?

@brampitoyo Noticed that Chrome Dev does exactly this in both forward and back directions. 

Pausing the forward or back gesture correctly orients the arrow affordances (this is incorrect in release Chrome) and matches what you see in Firefox desktop. The stack is also shown and can be selected by users. 

The stack is shown based on the context -- forward stack shows the forward stack, the back stack shows the back stack -- they don't mix it all into one stack. I prefer this behavior.

See the video I recorded: 
[signal-attachment-2020-05-17-124021.zip](https://github.com/mozilla-mobile/fenix/files/4640614/signal-attachment-2020-05-17-124021.zip)

**Comment by brampitoyo** [comment]
@yoasif Thanks for posting a video demo of Chrome Dev.

I think there’s a solution that may work in a majority of cases:

1. It doesn’t override Android system back button or gesture. It only supplements them with an additional “hold” action – which would already be familiar to Fennec and some Chrome users.
2. It doesn’t introduce any new button in the UI
3. It cleanly separates the back and forward stacks – which happens to be our Fennec implementation that @liuche and Engineering wanted to move forward with

The design works like this:

1. Back stack:
    * On Android 9 and below: long-press the Android system back button (identical to Fennec)
    * On Android 10 and above: swipe either left or right, then hold gesture. How long? For an identical amount of time as long-pressing the system back button.
2. Forward stack:
    * On all version: long-press the Forward (→) button in the main menu

![History Stacks@2x](https://user-images.githubusercontent.com/51007/82164679-da9f9880-9905-11ea-92f4-adc654fc6c9c.png)

@liuche @kbrosnan what do you think?

**Comment by yoasif** [comment]
> @yoasif Thanks for posting a video demo of Chrome Dev.
> 
> I think there’s a solution that may work in a majority of cases:
> 
>     1. It doesn’t override Android system back button or gesture. It only supplements them with an additional “hold” action – which would already be familiar to Fennec and some Chrome users.
> 
>     2. It doesn’t introduce any new button in the UI
> 
>     3. It cleanly separates the back and forward stacks – which happens to be our Fennec implementation that @liuche and Engineering wanted to move forward with
> 
> 
> The design works like this:
> 
>     1. Back stack:
>        
>        * On Android 9 and below: long-press the Android system back button (identical to Fennec)
>        * On Android 10 and above: swipe either left or right, then hold gesture. How long? For an identical amount of time as long-pressing the system back button.
> 
>     2. Forward stack:
>        
>        * On all version: long-press the Forward (→) button in the main menu

@brampitoyo I can't believe I didn't notice this before, but I just looked at Fennec again, and it seems that Fennec shows the back and forward stack on holding the Android back button.

Given that, I would propose a modification to your proposal to have a clean handoff from Fennec:

On all Android versions:

* Long press Android back button to show back and forward stack, with current page in boldface newest page to oldest (from top to bottom).
*  swipe either left or right, then hold gesture. How long? For an identical amount of time as long-pressing the system back button. This will show the corresponding back or forward stack based on the direction the user is gesturing.

My rationale here is that feels very weird to long press a menu item - it is not a metaphor I have seen anywhere in Android.

For the swipe gesture showing different stacks, it is more coherent with the way Chrome Dev does this, and it also just makes more sense, since holding the gesture shows the pages in the direction that the arrow shows. 

For the fact that the back *button* shows both stacks at once, we can call it convenience based on the fact that we're basically overloading a built in button in Android to provide a stack *period* and also continuity from Fennec -- and not requiring users to open a menu (a tap, a point and hold vs. a long press).

We also aren't forcing a possibly unnatural long press on "back" on any version of Android - gestures will be available on all versions of Android with the coherent stacks accessible via long gesture. 

What do you think @brampitoyo ?

**Comment by madb1lly** [comment]
Hi @yoasifm,

By "swiping" do you mean the system gesture navigation in Android 9+? In that case swiping forwards won't work on any phone which doesn't have this, which includes Android 9 phones which were upgraded from an earlier version. I know you said that long-pressing back could show back and forward stacks but this is a cludge and counter-intuitive.

About long-pressing the system back button to get the back stack - yes, @kbrosnan mentioned that earlier, please look above.

Long-pressing the back and forwards menu items to get the stack is how it works in Fennec because that is also how it works in Firefox Desktop. The UI paradigm is consistent between Firefox products, which is surely much more important than being consistent with an ever-changing Android UI changed at the whim of Google - try and follow that and you'll be playing catch-up forever.

Surely you can implement both swipe-and-old gesture _and_ long-press navigation methods into the product?

Cheers 🙂

**Comment by buttercookie42** [comment]
>     1. Back stack:
>        
>        * On Android 9 and below: long-press the Android system back button (identical to Fennec)
>        * On Android 10 and above: swipe either left or right, then hold gesture. How long? For an identical amount of time as long-pressing the system back button.

Also what the the following comment said, but specifically about this: I wouldn't hard code usage of the back button to an specific Android version, since you can disable gestures and return to the classic buttons even on Android 10 (and if you're using a third-party launcher this might actually happen automatically whether you want it or not).

**Comment by yoasif** [comment]
> Hi @yoasif,
> 
> By "swiping" do you mean the system gesture navigation in Android 9+? In that case swiping forwards won't work on any phone which doesn't have this, which includes Android 9 phones which were upgraded from an earlier version.

I think Fenix can do its own thing -- it isn't as if system gestures weren't available as third party software previously to Google introducing it, and it isn't as if we need to bring it to other apps, just Fenix. 

> I know you said that long-pressing back could show back and forward stacks but this is a cludge and counter-intuitive.

I suggested this because Fennec did this *and* because I think that long pressing a button (the forward button) inside of a menu feels very weird to me. It would make a lot more sense if I was pressing a forward button that was inside of a toolbar that I was long pressing.

> Long-pressing the back and forwards menu items to get the stack is how it works in Fennec because that is also how it works in Firefox Desktop. The UI paradigm is consistent between Firefox products, which is surely much more important than being consistent with an ever-changing Android UI changed at the whim of Google - try and follow that and you'll be playing catch-up forever.

Yes, but in Fennec, the back and forward button is present in the toolbar in tablet mode. In phone mode, no back button or forward button is drawn by Fennec, and it relies on the system back button to show both stacks.

I did just check Fennec, and long pressing the forward button inside of the menu *does* show the forward stack, and long pressing the back button inside of the menu shows the back stack.

So in that sense, the new behavior suggested by @brampitoyo is consistent in  a way with previous versions of Firefox - with the difference being that there is no back button drawn by Fenix in the toolbar like there was in Fennec (except in tablet mode). 

We don't yet have a tablet/large screen mode in Fenix, but if we did, I would suggest that Fennec had this right and that we should separate the stacks to the on screen buttons. 

Removing the forward stack might be seen as a regression by Fennec users (when using the system back button) and that is what I was suggesting we avoid. 

> 
> Surely you can implement both swipe-and-old gesture _and_ long-press navigation methods into the product?

Well, that is what I suggested. :)

**Comment by madb1lly** [comment]
> I think Fenix can do its own thing -- it isn't as if system gestures weren't available as third party software previously to Google introducing it, and it isn't as if we need to bring it to other apps, just Fenix.

Good idea 😀
> > I know you said that long-pressing back could show back and forward stacks but this is a cludge and counter-intuitive.
> 
> I suggested this because Fennec did this _and_ because I think that long pressing a button (the forward button) inside of a menu feels very weird to me. It would make a lot more sense if I was pressing a forward button that was inside of a toolbar that I was long pressing.

On my my Android 9 install of Fennec on a phone has capacitive back/home/switcher buttons, which originally ran Android 7 and has been upgraded to 9 I have the back and forward buttons in the menu and also the system back button. Each back button _only_ shows the back stack on long press and the forward button _only_ shows the forward stack on long press. If there are no back or forward stacks then the buttons in the menu are greyed out. You're saying that your Fennec install shows the whole forward/back stack when long-pressing back?
> > Long-pressing the back and forwards menu items to get the stack is how it works in Fennec because that is also how it works in Firefox Desktop. The UI paradigm is consistent between Firefox products, which is surely much more important than being consistent with an ever-changing Android UI changed at the whim of Google - try and follow that and you'll be playing catch-up forever.
> 
> Yes, but in Fennec, the back and forward button is present in the toolbar in tablet mode. In phone mode, no back button or forward button is drawn by Fennec, and it relies on the system back button to show both stacks.

In phone mode on my phone, as I said above, I have back and forward buttons in the menu as well as having the system back button. Maybe Fennec presents different buttons depending on the Android version or something which is why you're seeing different to me?

> I did just check Fennec, and long pressing the forward button inside of the menu _does_ show the forward stack, and long pressing the back button inside of the menu shows the back stack.

Yes that's how it works. Now I'm confused because that's not how you said it worked on your phone in the previous paragraph...

> So in that sense, the new behavior suggested by @brampitoyo is consistent in a way with previous versions of Firefox - with the difference being that there is no back button drawn by Fenix in the toolbar like there was in Fennec (except in tablet mode).

I do actually like what @brampitoyo was suggesting, I just didn't agree with my understanding of your interpretation of it.
 
> We don't yet have a tablet/large screen mode in Fenix, but if we did, I would suggest that Fennec had this right and that we should separate the stacks to the on screen buttons.

If by this you mean that Fenix copies Fennec then I agree 😀

> Removing the forward stack might be seen as a regression by Fennec users (when using the system back button) and that is what I was suggesting we avoid.

Excellent 😀

> > Surely you can implement both swipe-and-old gesture _and_ long-press navigation methods into the product?
> 
> Well, that is what I suggested. :)

I didn't realise that, I had read what you wrote differently, thanks for clarifying.

Cheers 🙂

**Comment by yoasif** [comment]
@madb1lly 
 
> You're saying that your Fennec install shows the whole forward/back stack when long-pressing back?

Yes, the system back menu, not the one in the menu. 

> In phone mode on my phone, as I said above, I have back and forward buttons in the menu as well as having the system back button. Maybe Fennec presents different buttons depending on the Android version or something which is why you're seeing different to me?

I am referring to the system back button, not the menu buttons.

The reason I am largely ignoring the menu buttons in Fennec is due to the fact that Fenix only has the forward button in the menu, coupled with the fact that it continues to feel like a strange interaction to me to long press a menu item. 

I think that if Fenix wants to have a long press action, it should be in a toolbar, not inside a menu, and since @brampitoyo hasn't proposed bringing back or forward buttons into a toolbar in phone mode, the Fennec back button method feels like the best path forward. 

Of course, we *could* add toolbar buttons in a toolbar like Opera, Edge and Samsung Internet do, but that doesn't feel like something that is being proposed here. 

My objection continues to be that long pressing an item in a menu feels like a very weird interaction.

**Comment by madb1lly** [comment]
Hi @yoasif,

Okay I've caught up with you now. Yes, when I press the system back button in Fennec I get the full history stack, forward and back. I suppose that would work fine in Fenix too.

About long-pressing a menu item: I think there are always going to be different opinions on how a UI should work and we're never going to find a solution that works for everyone. As an non-Mozilla employee, not even a "Mozillian" but as an enthusiastic user of Mozilla's products, from what I'm seeing in the Fenix issues discussions here on GitHub I'm concerned that some important features are not being taken because of concerns about UI. In these cases I think the project would benefit more from just implementing _something_ that works, even if it does not meet all the UI criteria. There are many other important features for Fennec migrants which I won't list here but are also suffering the same fate as this issue.

Back on topic - either what you suggest or @brampitoyo would work, so please let's get it implemented so we don't risk losing more people to Chrome during this transition period (I'm sure that some beta users have already left Fenix, although I hope they just went back to Fennec).

Cheers 🙂

**Comment by liuche** [comment]
The only action item for this SPIKE is to determine what engineering work will be necessary, and then it can be closed and summarized in this ticket, and @vesta0 can create a new issue.

**Comment by AmyYLee** [comment]
Seems like this no long needs UX-feedback. Removed tag. Please re-add if necessary

**Comment by person808** [comment]
So chrome does have a hidden setting that allows triggering the tab stack history with a gesture similar to what @brampitoyo described. However, it only works when the system gestures are disabled since it only triggers when swiping from the edge of the screen.

This is what chrome's gesture looks like after setting chrome://flags/#overscroll-history-navigation to "Enabled Navigation Sheet": https://streamable.com/uljnbl

**Comment by person808** [comment]
I'm looking into implementing a related swipe gesture in #12327. Swiping over the web page content likely requires A-C to expose a couple of APIs from GeckoView to allow us to detect if the web page can horizontally scroll. Assuming there are no issues with this (looking at you pull-to-refresh) we can implement add a swipe and hold gesture as part of #12327 or add it after.

From the Fenix side, #12174 would implement the glue code necessary to intercept swipes on web page content before `EngineView` gets a chance to handle them. All we would need to do is add a listener to detect swipes on `EngineView`, and trigger the gesture if the web page doesn't consume the touch events and can't scroll horizontally.

**Comment by brampitoyo** [comment]
@person808 The recommendation I wrote on https://github.com/mozilla-mobile/fenix/issues/12327#issuecomment-654566489 was identical to what I already wrote above, in this issue: https://github.com/mozilla-mobile/fenix/issues/1048#issuecomment-629891000

**Comment by person808** [comment]
I think this and #12327 are similar enough (gestures for navigating through tab history) that we should track them in one issue. I've laid out what needs to be done in #12327.

**Comment by yoasif** [comment]
@person808 It is similar for gesture based navigation, sure, but not in the non-gesture based case - see my comment https://github.com/mozilla-mobile/fenix/issues/1048#issuecomment-629902570.

In terms of prioritization, do most Fenix users have gesture based controls enabled in their OS, or are they using buttons? Would be helpful to know in order to get the design of the stack (I think this is going to be shared anyway) right.

**Comment by person808** [comment]
@yoasif You're right. On second thought they differ enough that its probably better to keep them separate for now.

>In terms of prioritization, do most Fenix users have gesture based controls enabled in their OS, or are they using buttons? Would be helpful to know in order to get the design of the stack (I think this is going to be shared anyway) right.

Honestly I'm not sure. But its probably a safe assumption that more people will be using gestures as they buy phones with Android 10. But the stack design itself likely wouldn't change, only the way to trigger the it would.

@brampitoyo Here's what I think we can do for this issue:

- Can be done now:
    - When the forward button (inside the main menu) is long-pressed, show history stack (needs A-C support for long-pressing a menu item)
    - When the back button (provided by Android) is long-pressed, show history stack
- Needs UX work:
    - Figure out alternative way to show history stack when the back button is unavailable due to system gestures being enabled (we can't detect a swipe and hold on the system back gesture)

**Comment by brampitoyo** [comment]
@person808 Thanks heaps for this!

Even if we can’t detect swipe-and-hold system back gesture, users can still get access to the history stack via tap-and-hold the forward button inside the main menu. It’s not perfect, but it means that Android 10+ users won’t miss out on this feature.

**Comment by Scripterr** [comment]
Hello @person808 @brampitoyo , if I'm understanding this point correctly from above statement

> Can be done now:
> When the forward button (inside the main menu) is long-pressed, show history stack (Blocked by A-C support for long-pressing a menu item)
> When the back button (provided by Android) is long-pressed, show history stack

That means we are going to have two history stack one each for forward and back but why can't we have a unified from just like in fennec and it will be much easier to navigate through tabs from one place rather than going through menu and then pressing forward button to show forward stack.

 Plus many users wouldn't even know that there is a forward stack and it's more inconvenient that you have to jump through hoops to get forward stack and users may not be using because off it. I'm not a ui person so can't say it's perfect or problem free but if there is any UX issue with it then please continue with original idea , I only want to express my thoughts on this. I would be fine with anyway it would be implemented.
Thank you.

**Comment by brampitoyo** [comment]
@Scripterr As far as I’m aware, we are only going to have a single history stack, not two.

Let’s say that you’ve visited a few webpages in this order: 1, 2, 3, 4, 5. Then you hit the Back button twice, so you’re in page 3. Now you enter the history stack – either using the back or the forward button. Doesn’t matter which, because both will work.

What will this history stack show? It will display:

- Page 1
- Page 2
- **Page 3** ← You are here
- Page 4
- Page 5

Whichever way you want to go, back or forward, you can do so using this “unified” history stack.

Does that sound like a good explanation?

@person808 can confirm.

**Comment by person808** [comment]
@brampitoyo I can definitely do that

**Comment by madb1lly** [comment]
Lovely UI mockup @brampitoyo 😃

**Comment by pocmo** [comment]
@brampitoyo If the long press on back is no longer an officially supported pattern (by the way also before the switch to gestures, see below[1]) then wouldn't we want to think of alternatives to show this UI? Is it time to think about having a back button next to the forward button inside the menu? Long-pressing the forward button to show UI so that I can navigate back to a specific page is a bit awkward? 🤔 

[1] Fennec contains this workaround since Android already stopped calling onBackLongPressed with version 7 (N):
https://searchfox.org/mozilla-esr68/source/mobile/android/base/java/org/mozilla/gecko/BrowserApp.java#584-613

**Comment by person808** [comment]
For clarification: the standard way to handle long pressing the back button is only broken on Android 7. I haven't thoroughly tested but it works on Android 8 and Android 10 for me. In addition, the only [issue](https://issuetracker.google.com/issues/37106088) regarding broken long press for the back button on Google's issue tracker only mentions Android 7 and is now closed. I have ported the workaround linked above and it seems to work well for Android 7.

**Comment by brampitoyo** [comment]
@pocmo I didn’t know that long press on back is no longer an officially supported pattern. I’m running Android 10. When setting gesture navigation to 3-button, I saw that Chrome still supported long-press to access tab history.

**Comment by pocmo** [comment]
@brampitoyo Okay, "no longer officially supported" was not really the right choice of words. But we know that:
* It is broken on Android 7ish (There's a workaround we implemented in Fennec though)
* It is not working with gesture navigation (Android 10+) - There's probably no new phone coming out that will sacrifice parts of it screen for 2/3 button navigation?
* It is pretty much undiscoverable (Has any app except browsers ever used it?)

So I was wondering if we should find a new future proof pattern? :) Does Chrome do anything new on Android 10+ with gesture navigation?

**Comment by boek** [comment]
@pocmo re: new patterns. We couldn't find anything in Chrome

**Comment by brampitoyo** [comment]
@pocmo Yes. I agree with @boek.

Chrome (and Chrome Canary with all UI flags switched ON, just in case) doesn’t do anything new with gesture navigation to access history.

On Android 11, I’m not so sure. Can anybody check?

**Comment by Augures01** [comment]
When I long press (physical) back button, history always opens twice
Samsung S6 Edge, Android 7.0
Nightly 200723 18:05

**Comment by yoasif** [comment]
Wait, this is live now? I didn't even realize it!

**Comment by person808** [comment]
@Augures01 could you file that as a new issue? Thanks!

**Comment by opusforlife2** [comment]
@yoasif 
> Wait, this is live now? I didn't even realize it!

@person808 is very sneaky. Merged both tab history and swipe but didn't have the linked issues closed. 🤣

**Comment by madb1lly** [comment]
> 
> 
> @yoasif
> 
> > Wait, this is live now? I didn't even realize it!
> 
> @person808 is very sneaky. Merged both tab history and swipe but didn't have the linked issues closed. 🤣
Exactly why we need a newsletter, blog, or a generally better way of informing people what's new in nightly #11719.

Cheers 🙂

**Comment by ShatteredPixel** [comment]
@madb1lly The reason why this issue isn't closed is because it isn't yet verified by the QA team of it is working as intended or not or are there any bugs with it? As you can see this issue contains the label `eng:qa:needed`. Nighties change too much at any given time and some features are sometimes rolled back. It'd hard to keep track of everything. And instead when the beta and release builds are released, the change-log is collected and displayed in the release notes in the Paly Store and the Fenix's GitHub release page. You can keep track of the Nightly changes in PR section and see which one are being merged and which one have the `eng:qa:needed` label, those features would have been implemented. That's the only way how it works now. It is exactly for this reason of Nightly changing a lot I think the change-log isn't pushed to the general users and even to the advanced users. You can see the feature being rolled back like "pull to refresh" and "dynamic toolbar" due to severe bugs with them.

**Comment by ebalazs-sv** [comment]
Hi!
I had tested this issue on Nightly 200727 Build #22090608 GV 80.0a1 from 7/27 with the following devices:
- Google Pixel 4 XL (Android 11)
- Samsung Note 10 (Android 10)
- Google Pixel 2 (Android 9);
- Motorola Moto G6 (Android 8);
- Huawei P9 Lite (Android 7);
- Nexus 5 (Android 6.0.1);
- Huawei MediaPad M2 (Android 5.1.1);

and the recent history pop-up is displayed in the following scenarios:
- long tapping the device back button;
- long tapping the forward app button,

in landscape mode, portrait mode, dark theme, light theme, private browsing, normal browsing.

I had filed the following related issues:
https://github.com/mozilla-mobile/fenix/issues/12972
https://github.com/mozilla-mobile/fenix/issues/12973
https://github.com/mozilla-mobile/fenix/issues/12974
https://github.com/mozilla-mobile/fenix/issues/12977
https://github.com/mozilla-mobile/fenix/issues/12978

I will remove the qa:needed label and close this issue.

**Comment by madb1lly** [comment]
Hi @ShatteredPixel,
All of that I know, thanks for reminding me. Still, there is a Nightly newsletter for Firefox desktop, so it's odd there can't be one for Mobile too.
Anyway, that's off topic.
Cheers 🙂

**Comment by brampitoyo** [comment]
@pocmo I think we may add a back button in the main menu navbar, as you originally suggested on https://github.com/mozilla-mobile/fenix/issues/1048#issuecomment-658818520. It may be the only possible solution for this problem.

GitHub issue #12926.

**Comment by brampitoyo** [comment]
@pocmo @jonalmeida @person808 I’ve found something interesting earlier today. It’s a Chrome flag which shows a history stack on a swipe-and-hold gesture.

This was exactly the behaviour I recommended on https://github.com/mozilla-mobile/fenix/issues/1048#issuecomment-629891000. My spec said that it should work when Gesture Navigation is ON. Unfortunately, Gesture Navigation is precisely what’s preventing swipe-and-hold to work!

Steps to reproduce:

- Set gestures to 2-button navigation
- Open Chrome
- Type `chrome://flags`
- Set `overscroll-history-navigation` to `Enabled Navigation sheet`
- Relaunch Chrome
- In one tab, open a set of pages in succession
- Swipe from the left, then hold

![swipe-and-hold-history-stack](https://user-images.githubusercontent.com/51007/89256245-c8690700-d677-11ea-90c0-b9f0bd5f1b78.gif)

**Comment by Shaurya-Kalia** [comment]
Assigning it to a gesture is not the best solution in my view since different android versions and skins have different gesture implementations. The best way to go about it would be to have a dedicated back/forward buttons

**Comment by madb1lly** [comment]
Hi all,

Aren't there some configuration settings in Android which can be queried to determine the hardware UI which is being used? This would make it less vague then basing it on Android versions. The hardware UI setting would then determine whether the extra back button in the main menu or the swipe and hold gesture is provided.

E.g., in my case with a Nokia 8, it has hardware capacitative buttons so I don't need either the extra back button or the swipe and hold gesture (the latter wouldn't work anyway).

Cheers 🙂

**Comment by cadeyrn** [comment]
@madb1lly see https://github.com/mozilla-mobile/fenix/issues/12926#issuecomment-663826740. The back button in the menu is not only useful for people with enabled gestures navigation.

**Comment by jonalmeida** [comment]
@brampitoyo that's pretty interesting (and an unfortunate case for Chrome)!

Based on other comments about potential conflicts with gesture navigation on different Android flavours like Samsung's TouchWiz, I wonder if we can take a page from [Opera Touch](https://play.google.com/store/apps/details?id=com.opera.touch) and create a "gesture mode" that enables all these features by user choice with a small onboarding card of sorts.

While that's quite a grandiose idea, we could possibly just start with an onboarding card to begin with (similar to our dark/light theme card).

Just throwing some ideas out there. :)

**Comment by brampitoyo** [comment]
Hearing more suggestions from all of you @Shaurya-Kalia @madb1lly @cadeyrn @jonalmeida made me believe that a permanent back button sounds like the most sensible idea to pursue in the short-term.

- Even if we have gestures compatible with many versions of Android, different skins could simply choose not to respect it.
- We can’t expect all phones to come with hardware or software back buttons, but if they do have it, can we still respect the tap-and-hold gesture?
- Doing hardware/software detection to disable/enable features sound like it could lead to more difficult maintenance and more corner cases.
- This time, we decided not to implement any gesture to access the history stack. However, for every new gesture we implement in the future, we should expose and make them controllable from within Settings.

**Comment by person808** [comment]
> Doing hardware/software detection to disable/enable features sound like it could lead to more difficult maintenance and more corner cases.

100% agree. Its not as simple as doing something like `isSystemGesturesEnabled()` (I wish it was). We can use `WindowInsets` to detect the area reserved for system gestures, but that API comes with its own quirks.

As an engineering side note: If we do want to go down the route of adding fancier gestures, we might be able to extend/use [`SwipeGestureLayout`](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/browser/SwipeGestureLayout.kt). This provides the swipe detection for swiping to switch tabs, but its actually set up so it can detect swipes anywhere on the browser screen.

**Comment by yoasif** [comment]
>     * We can’t expect all phones to come with hardware or software back buttons, but if they do have it, can we still respect the tap-and-hold gesture?

@brampitoyo We just added it - are we really talking about possibly unsupporting it already, or am I misinterpreting this comment?

**Comment by M86xKC** [comment]
There are many other options, but most are not as discoverable as long-pressing on a back button.

Here are some Ideas: 

1. Long-pressing the tab icon (next to the address) already shows (New tab, Net private tab then close bar), so maybe add it to the list there as back history ?. (most discoverable IMO)

2. Long-pressing the three dot does nothing as of now, so that can also be used (not discoverable).

3. Using gestures, as @person808 said there could be an option to swipe vertically on the address bar (up/down depending on where you have placed the address bar) to access the history stack. Maybe add a little bit of delay (more like swipe and hold) to prevent accidentally triggering it while scrolling.

**Comment by Dunexus** [comment]
> 3. Using gestures, as @person808 said there could be an option to swipe vertically on the address bar (up/down depending on where you have placed the address bar) to access the history stack. Maybe add a little bit of delay (more like swipe and hold) to prevent accidentally triggering it while scrolling.

This gesture is currently considered to be used to open tab tray, see https://github.com/mozilla-mobile/fenix/issues/11862

**Comment by madb1lly** [comment]
> 
> 
> > * We can’t expect all phones to come with hardware or software back buttons, but if they do have it, can we still respect the tap-and-hold gesture?
> 
> @brampitoyo We just added it - are we really talking about possibly unsupporting it already, or am I misinterpreting this comment?
Hi @brampitoyo and @yoasif,

I'm not exactly sure what is meant by the "tap-and-hold" gesture yet I hope that the conclusion is to keep the long-press on _any_ back button to show the history stack, whether that be an existing back hardware button or a new back button in the menu.

Cheers 🙂

**Comment by jonalmeida** [comment]
> > > * We can’t expect all phones to come with hardware or software back buttons, but if they do have it, can we still respect the tap-and-hold gesture?
> > 
> > 
> > @brampitoyo We just added it - are we really talking about possibly unsupporting it already, or am I misinterpreting this comment?
> > Hi @brampitoyo and @yoasif,
> 
> I'm not exactly sure what is meant by the "tap-and-hold" gesture yet I hope that the conclusion is to keep the long-press on _any_ back button to show the history stack, whether that be an existing back hardware button or a new back button in the menu.
> 
> Cheers 🙂

I believe @brampitoyo is referring to this tap-to-hold gesture from a related ticket: https://github.com/mozilla-mobile/fenix/issues/12926#issuecomment-665417017

**Comment by madb1lly** [comment]
Hi @jonalmeida,
Really? That looks more like hold and swipe than tap and hold... another reason to get clarification from @brampitoyo 😆
Cheers

**Comment by M86xKC** [comment]
> 11862

We can still use swipe up and hold, no?

**Comment by brampitoyo** [comment]
@madb1lly @M86xKC Unfortunately, we can’t use swipe-and-hold.

We do have the ability to use hold-and-swipe – which a reverse of swipe-and-hold. It’s when you tap on the side of the screen and hold it for a moment. The system lets you know that something is happening, by opening a little bit of a sheet. If you then swipe to the centre of the screen, the screen expands.

This gesture, while smart, has zero affordance. There’s no indicator which tells you “this is where you can tap-hold-and-swipe”, because putting a visual like that would distract people away from the website (main content) they’re viewing. If you don’t know exactly how to trigger it, it’s not discoverable.

Plus, Android designed this gesture specifically to access the three-line menu, and not for any other element.

For those reasons, we decided against adopting this gesture in favour of simply inserting an unmistakeable ← icon inside the main menu.

**Comment by chriscjs** [comment]
This works on my Pixel 3a running stock but not on Nexus 6p running Lineage OS 17.1.  Both phones have same version of Firefox beta (8-18).  So it obviously is possible.

**Comment by person808** [comment]
> This works on my Pixel 3a running stock but not on Nexus 6p running Lineage OS 17.1. Both phones have same version of Firefox beta (8-18). So it obviously is possible.

@chriscjs What is "this" referring to? Implementations of gesture navigation vary between manufacturer and Android versions so relying on a swipe and hold gesture is not an option here. FWIW, swipe and hold does not work on my Pixel 2 XL also on stock Android. If a swipe and hold does work its simply a nice thing that the manufacturer/ROM developer has implemented.

**Comment by chriscjs** [comment]
Long pressing back button at bottom of screen, which is what I thought this issue was about.  Also works in Chrome on Nexus 6p.

**Comment by person808** [comment]
Ah ok. The long press back button is already implemented which is why this was closed :) If it doesn't work then its probably a device/manufacturer specific issue and you should open a new issue to track it.
